plugin/Categorifier/Core.hs

```diff
@@ -492,6 +499,50 @@ applyCategorify ::
   (Plugins.Type -> Plugins.CoreExpr -> CategoryStack Plugins.CoreExpr) ->
   [Plugins.CoreExpr] ->
   IO (Maybe Plugins.CoreExpr)
+#if MIN_VERSION_GLASGOW_HASKELL(9, 0, 0, 0)
+applyCategorify convert hierarchyOptions defer dflags uniqS f = \case
+  (Plugins.Type _mult : Plugins.Type cat : Plugins.Type a : Plugins.Type b : calls : function : extraArgs) ->
+    if null extraArgs
+      then
+        pure
+          <$> runStack
+            hierarchyOptions
+            ((\fn -> fn cat a b calls) <$> defer)
+            dflags
+            uniqS
+            calls
+            function
+            (f cat)
+      else
+        throwIOAsException
+          ( \args ->
+              [fmt|
+Categorifier: GHC somehow called `{Plugins.getOccString convert}` with too many
+           arguments. This shouldn't have gotten past the typechecker:
+           {Plugins.showPpr dflags args}
+|]
+          )
+          extraArgs
+  -- The next few cases are partial applications of `categorify` that we ignore and hope finish
+  -- application later (see `applyApply`).
+  [Plugins.Type _mult, Plugins.Type _cat, Plugins.Type _a, Plugins.Type _b, _callStack] -> pure Nothing
+  [Plugins.Type _mult, Plugins.Type _cat, Plugins.Type _a, Plugins.Type _b] -> pure Nothing
+  [Plugins.Type _mult, Plugins.Type _cat, Plugins.Type _a] -> pure Nothing
+  [Plugins.Type _mult, Plugins.Type _cat] -> pure Nothing
+  [Plugins.Type _mult] -> pure Nothing
+  [] -> pure Nothing
```

Comment:
From this to line 544, it's the same as non-linear version. we had better minimize conditional compilation part.

Reply:
This can be difficult with Ormolu. They've changed the way `CPP` is handled a few times, but I _think_ the current approach is that CPP directives are treated as comments, and then the module is formatted.

Generally, this means that putting directives around top-level definitions works just fine. Other cases may work, but are more fragile. Here, shrinking the directives to only cover some of the `case` alternatives causes a parse error. I'm not sure exactly why, as it seems like treating the directives as comments should allow that to work. But I'm guessing there's some subtlety in the CPP handling that I'm missing.

Reply:
I see. yeah. I was often bitten by ormolu and CPP incompatibility. If not easily fixable, please disregard my suggestion.